src/dvi.rs

```diff
@@ -102,39 +116,127 @@ impl<Channels: DmaChannelList> DviInst<Channels> {
+    /// available, so the display list can be advanced.
     #[link_section = ".data"]
-    fn render(&mut self) {
-        if let Some(y) = self
-            .timing_state
-            .v_scanline_index(&self.timing, TMDS_PIPELINE_SLACK)
-        {
+    fn render(&mut self, y: u32, available: bool) {
+        let buf_ix = y as usize % N_TMDS_BUFFERS;
+        let line_size = self.timing.horizontal_words() as usize * N_CHANNELS;
+        let line_start = line_size * buf_ix;
+        let tmds_slice = &mut self.tmds_buf[line_start..][..line_size];
+        self.scan_render.render_scanline(tmds_slice, y, available);
+    }
+
+    /// Schedule the rendering of a line buffer.
+    ///
+    /// The line buffers are rendered outside the main interrupt handler,
+    /// striped across both cores.
+    #[link_section = ".data"]
+    fn schedule_line_render(&mut self) {
+        let offset = TMDS_PIPELINE_SLACK + (N_LINE_BUFS * VERTICAL_REPEAT) as u32;
+        if let Some(y) = self.timing_state.v_scanline_index(&self.timing, offset) {
             if y % VERTICAL_REPEAT as u32 == 0 {
                 let y = y / VERTICAL_REPEAT as u32;
-                let buf_ix = y as usize % N_TMDS_BUFFERS;
-                let line_size = self.timing.horizontal_words() as usize * N_CHANNELS;
-                let line_start = line_size * buf_ix;
-                let tmds_slice = &mut self.tmds_buf[line_start..][..line_size];
-                self.scan_render.render_scanline(tmds_slice, y);
+                self.scan_render.schedule_line_render(y);
             }
         }
     }
 }
 
+/// We dedicate core 1 to running the primary video interrupt and also
+/// background rendering tasks.
+#[link_section = ".data"]
+pub fn core1_main<PIO, SliceId, Pos, Neg, RedPos, RedNeg, GreenPos, GreenNeg, BluePos, BlueNeg>(
+    mut serializer: DviSerializer<
+        PIO,
+        SliceId,
+        Pos,
+        Neg,
+        RedPos,
+        RedNeg,
+        GreenPos,
+        GreenNeg,
+        BluePos,
+        BlueNeg,
+    >,
+) -> !
+where
+    PIO: pio::PIOExt,
+    SliceId: pwm::SliceId,
+    Pos: PinId + bank0::BankPinId + ValidPwmOutputPin<SliceId, pwm::A>,
+    Neg: PinId + bank0::BankPinId + ValidPwmOutputPin<SliceId, pwm::B>,
+    RedPos: PinId + bank0::BankPinId,
+    RedNeg: PinId + bank0::BankPinId,
+    GreenPos: PinId + bank0::BankPinId,
+    GreenNeg: PinId + bank0::BankPinId,
+    BluePos: PinId + bank0::BankPinId,
+    BlueNeg: PinId + bank0::BankPinId,
+{
+    unsafe {
+        NVIC::unmask(Interrupt::DMA_IRQ_0);
+    }
+    serializer.wait_fifos_full();
+    serializer.enable();
+    loop {
+        let line_ix = CORE1_QUEUE.peek_blocking();
+        // Safety: exclusive access to the line buffer is granted
+        // when the render is scheduled to a core.
+        unsafe { render_line(line_ix) };
+        CORE1_QUEUE.remove();
+    }
+}
+
 #[link_section = ".data"]
 #[interrupt]
 fn DMA_IRQ_0() {
```

Comment:
Just to document https://github.com/DusterTheFirst/pico-dvi-rs/pull/12#discussion_r1305013352
```suggestion
/// Called on core 1 every scan line by the DMA controller
#[link_section = ".data"]
#[interrupt]
fn DMA_IRQ_0() {
```